src/lib/address_resolve/AddressResolve_DefaultImpl.h

```diff
@@ -26,13 +26,44 @@ namespace chip {
 namespace AddressResolve {
 namespace Impl {
 
+constexpr uint8_t kNodeLookupResultsLen = CHIP_CONFIG_MDNS_RESOLVE_LOOKUP_RESULTS;
+
 enum class NodeLookupResult
 {
     kKeepSearching, // keep the current search active
     kLookupError,   // final status: error
     kLookupSuccess, // final status: success
 };
 
+struct NodeLookupResults
+{
+    ResolveResult results[kNodeLookupResultsLen] = {};
+    uint8_t count                                = 0; // number of valid ResolveResult
+    uint8_t consumed                             = 0; // number of already read ResolveResult
+
+    bool UpdateResults(const ResolveResult & result, Dnssd::IPAddressSorter::IpScore score);
+
+    bool HasValidResult() const { return count > consumed; }
+
+    ResolveResult ConsumeResult()
+    {
+        VerifyOrDie(HasValidResult());
+        return results[consumed++];
```

Comment:
This part is not making sense to me, unless I am missing something.

Let's say we get two results, A and B.  Array is [A, B].  We consume A then a new result C comes in with a better score.  The array ends up as [C, A, B], `consumed` is set to 1, so the next time we will try A again, then try B, and never try C.

Fundamentally, since new results can get inserted in between existing possibly-consumed ones with our scoring setup a single "consumed" counter doesn't seem to be able to represent the state of things.

I see two possible solutions offhand:
1. ConsumeResult in fact moves things over by one and reduces `count`.
2. We change the ordering of things in the array so smaller scores come first.  That makes it cheap to add a new thing with a new high score (no shifting needed), and cheap to pop off the thing with the best score (just decrement count), but makes adding a new low-scoring thing more expensive (need to shift, and need to be careful to not shift out high-scoring things).

Either way, `consumed` would go away.

Looking at the tests, they don't seem to test these "interleave consumption with addition" scenarios, and probably should.

Reply:
As discussed on slack I have added a check in `TryNextResult` to make it clearer that the lookup handle is not expected to be active when it is called. As you said that would result into all sort of weird results.